# Primeira Avaliação Parcial: Avaliação Individual

Neste módulo trabalhamos com os recursos da __NumPy__ para processar dados de tipo único de forma eficiente. Um exemplo deste tipo de dados são as imagens que podemos ser armazenadas como _ndarrays_ __NumPy__. O código a seguir gera uma imagem colorida, de 1000 por 1000 _pixels_, armazenada no formato __RGB__. Isto significa que, de cada _pixel_ armazenamos a intensidade da cor dos canais __R__ (_red_ ou vermelho), __G__ (green ou verde) e __B__ (blue o azul).  A intensidade da cor é dada por um inteiro de 8 _bits_ sem sinal.

In [1]:
# importado a biblioteca numpy
import numpy as np
print(np.__version__)


1.24.3


In [2]:
# Definindo a altura e a largura da imagem que represnearão:
height = 1000  # altura ou o número de linhas da imagem
width = 1000   # largura ou o número de colunas da imagem

In [3]:
# gerando uma imagem aleatória de 1000 x 1000 pixels
random_image = np.random.randint(0, 256, size=(height, width, 3), dtype=np.uint8)
print("Veja o forma da matriz criada: {}".format(random_image.shape))
print("Veja como fica a cor de um pixel: {} ...".format(random_image[0, 0]))
print("Ou a componente vermelha dos 3 primeiro pixels da segunda linha: {} ...".format(random_image[1, 0:3, 0]))
print("Ou a componente verde dos 3 últimos pixels da segunda coluna: {} ...".format(random_image[-3:, 1, 1]))

Veja o forma da matriz criada: (1000, 1000, 3)
Veja como fica a cor de um pixel: [ 51 138 231] ...
Ou a componente vermelha dos 3 primeiro pixels da segunda linha: [233 222  39] ...
Ou a componente verde dos 3 últimos pixels da segunda coluna: [188   7 200] ...


## Exercício 1

Crie uma função para a qual passamos como parâmetros a altura e a largura e retorna uma imagem, gerada de forma aleatória, em formato __RGB__, com estas dimensões. A imagem deve ser representada utilizando um _ndarray_ como o do exemplo anterior. Se a altura ou a largura não forme passadas como a imagem deve ser gerada com 720 linhas e 1280 colunas. 

In [4]:
# Implementar aqui
def geradorImagens(h = 720, w = 1280):
    return np.random.randint(low=0,high=256,size=(h,w,3),dtype=np.uint8)

In [5]:
#teste
print(geradorImagens().shape) # deve retornar (720, 1280, 3)
print(geradorImagens(10).shape) # deve retornar (10, 1280, 3)
print(geradorImagens(10,10).shape) # deve retornar (10, 10, 3)
print(geradorImagens(5,5)) # deve retornar uma matriz 5x5x3

(720, 1280, 3)
(10, 1280, 3)
(10, 10, 3)
[[[ 55 200 252]
  [ 79 210 244]
  [ 34 152 178]
  [  5 220 206]
  [244 163  18]]

 [[239  31 184]
  [  9  43 162]
  [246 229 120]
  [ 43  66  44]
  [ 68 159  21]]

 [[ 20 230 199]
  [129  26  51]
  [220 150 224]
  [ 83 105 197]
  [ 42  66  89]]

 [[174 125 226]
  [198 144 136]
  [181   1  37]
  [145 122 189]
  [217 121 210]]

 [[231 132 205]
  [ 44  75  33]
  [124 209 214]
  [ 43 246 235]
  [ 18   0 214]]]


Para simplificar o tratamento durante o processamento das imagens, frequentemente elas são convertidas e imagens em tons de cinza. Estas imagens contem, para cada pixel, uma única intensidade de cor representada como um inteiro de 8 _bits_ sem sinal. Para converter uma imagem __RGB__ em uma imagem em tons de cinza podemos utilizar dois métodos diferentes. 

1. Calculamos a intensidade da cor de cada _pixel_ como a média das intensidades dos canais __RGB_
2. Os valores __RGB__ são convertidos para tons de cinza usando a fórmula __NTSC__: $ 0.299 \times \text{Vermelho} + 0.587 \times \text{Verde} + 0.114 \times \text{Azul} $. Esta fórmula representa de perto a percepção relativa da pessoa média sobre o brilho da luz vermelha, verde e azul.

## Exercício 2

Crie uma função que recebe uma imagem __RGB__ e converte a mesma em tons de cinza. A função deve retornar então a nova imagem. Esta função. 
* Deve verificar se o argumento de entrada é de fato uma imagem __RGB__ com base no tipo e no atributo ``shape`` do _ndarray_. Casso o argumento esteja errado deve-se lançar uma exceção.
* A função recebe como parâmetro de entrada a forma de conversão a ser utilizada: media dos canais ou __NTSC__. Por padrão o algoritmo de utilizado deve ser o __NTSC__.
* A função deve utilizas as _ufunc_ e operações definidas na __NumPy_ de forma a minimizar o uso de laços no processamento dos _ndarrays_.  

In [6]:
#implementar aqui
def rgb2gray(array,forma = "ntsc"):
    if array.shape[2] != 3:
        raise Exception("MATRIZ COM FORMATO INVALÍDO")
    
    if forma.lower() == "ntsc":
        return (0.299*array[:,:,0] + 0.587*array[:,:,1] + 0.114*array[:,:,2]).astype(np.uint8)
    
    if forma.lower() == "mean":
        return np.mean(array, axis=2, dtype=np.uint8)
        

In [7]:

teste = geradorImagens(2,2)
print(teste)



[[[ 11 251 241]
  [227  69 119]]

 [[ 91 194  35]
  [159  71 161]]]


In [8]:
ntsc = rgb2gray(teste)
print(ntsc)
print()
mean = rgb2gray(teste,"mean")
print(mean)

[[178 121]
 [145 107]]

[[82 53]
 [21 45]]


# Exercício 3 (Opcional, valendo ponto extra)

Refaça a implementação da função anterior utilizando estruturas de repetição e implemente um teste de demonstre, usando uma imagem __RGB__ gerada de forma aleatória de 10.000 linhas por 10.000 colunas, o ganho de desempenho quando utilizamos as _ufunc_ e os operadores de __NumPy__. 

In [9]:
def rgb2grayFor(array,forma = "ntsc"):
    if array.shape[2] != 3:
        raise Exception("MATRIZ COM FORMATO INVALÍDO")
    

    h, w, _ = array.shape

    res = np.zeros(shape = (h,w))

    if forma.lower() == "ntsc":
        for i in range(h):
            for j in range(w):
                    res[i,j] = 0.299*array[i,j,0] + 0.587*array[i,j,1] + 0.114*array[i,j,2]
        
        return res.astype(np.uint8)
                    
    
    if forma.lower() == "mean":
        for i in range(h):
            for j in range(w):
                for k in range(3):
                    res[i,j]+=array[i,j,k]
        
        for i in range(h):
            for j in range(w):
                res[i,j] = res[i,j].astype(np.uint8)/3
        
        return res.astype(np.uint8)

Verificando resultados

In [10]:
print("resultado com ufuncs")
ntsc = rgb2gray(teste)
print(ntsc,ntsc.dtype)
print()
mean = rgb2gray(teste,"mean")
print(mean,mean.dtype)

print("\nresultado com laço for")
ntsc = rgb2grayFor(teste)
print(ntsc,ntsc.dtype)
print()
mean = rgb2grayFor(teste,"mean")
print(mean,mean.dtype)

resultado com ufuncs
[[178 121]
 [145 107]] uint8

[[82 53]
 [21 45]] uint8

resultado com laço for
[[178 121]
 [145 107]] uint8

[[82 53]
 [21 45]] uint8


teste de performace

In [12]:
img = geradorImagens(10000,10000)

print("resultado da implementação utilizando ufuncs")
%timeit rgb2gray(img)
%timeit rgb2gray(img,"mean")
print("\n\n resultado da implementação utilizando laço for")
%timeit -n 1 -r 1 rgb2grayFor(img)
%timeit -n 1 -r 1 rgb2grayFor(img,"mean")

resultado da implementação utilizando ufuncs
1.23 s ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
883 ms ± 6.62 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


 resultado da implementação utilizando laço for
7min 32s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
6min 20s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


resultado da implementação utilizando ufuncs
1.23 s ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
883 ms ± 6.62 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)



resultado da implementação utilizando laço for
7min 32s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
6min 20s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)

## Respostas
Faça suas implementações neste notebook e envie  mesmo via __Moodle__ até o final do prazo. 